plugins/newrelic/src/api/index.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { createApiRef, DiscoveryApi } from '@backstage/core';
+
+export type NewRelicApplication = {
+  id: number;
+  application_summary: NewRelicApplicationSummary;
+  name: string;
+  language: string;
+  health_status: string;
+  reporting: boolean;
+  settings: NewRelicApplicationSettings;
+  links?: NewRelicApplicationLinks;
+};
+
+export type NewRelicApplicationSummary = {
+  apdex_score: number;
+  error_rate: number;
+  host_count: number;
+  instance_count: number;
+  response_time: number;
+  throughput: number;
+};
+
+export type NewRelicApplicationSettings = {
+  app_apdex_threshold: number;
+  end_user_apdex_threshold: number;
+  enable_real_user_monitoring: boolean;
+  use_server_side_config: boolean;
+};
+
+export type NewRelicApplicationLinks = {
+  application_instances: Array<any>;
+  servers: Array<any>;
+  application_hosts: Array<any>;
+};
+
+export type NewRelicApplications = {
+  applications: NewRelicApplication[];
+};
+
+export const newRelicApiRef = createApiRef<NewRelicApi>({
+  id: 'plugin.newrelic.service',
+  description: 'Used by the NewRelic plugin to make requests',
+});
+
+const DEFAULT_PROXY_PATH_BASE = '/newrelic';
+
+type Options = {
+  discoveryApi: DiscoveryApi;
+  /**
+   * Path to use for requests via the proxy, defaults to /newrelic
+   */
+  proxyPathBase?: string;
+};
+
+export interface NewRelicApi {
+  getApplications(): Promise<NewRelicApplications>;
+}
+
+export class NewRelicClient implements NewRelicApi {
+  private readonly discoveryApi: DiscoveryApi;
+  private readonly proxyPathBase: string;
+
+  constructor(options: Options) {
+    this.discoveryApi = options.discoveryApi;
+    this.proxyPathBase = options.proxyPathBase ?? DEFAULT_PROXY_PATH_BASE;
+  }
+
+  async getApplications(): Promise<NewRelicApplications> {
+    const url = await this.getApiUrl('apm', 'applications.json');
+    const response = await fetch(url);
+    const responseJson = await response.json();
+
+    if (response.status !== 200) {
+      throw new Error(responseJson?.error?.title || response.statusText);
+    }
```

Comment:
bit of a nit, but `.json()` explodes if it wasn't a 2xx response with data i think, so you may want to do this check right after the fetch instead (or twice, in different ways, whichever is best)

Reply:
Then again maybe it's ok the way you have it now, if it's known that the api always returns a json body no matter what, and it's fine to explode on the very unexpected corner case

Reply:
That's a good call-out; if we're going to be friendlier with error messages, we may as well say so when New Relic is hard down too.

Reply:
Probably good to add a little extra context in any event.

**When `api.newrelic.com` is blackhole'd**:
![Screenshot 2020-10-15 at 12 28 32](https://user-images.githubusercontent.com/3496491/96112857-924fcb00-0ee3-11eb-8114-bc7e1433ca96.png)

**When the key is invalid**:
![Screenshot 2020-10-15 at 12 36 00](https://user-images.githubusercontent.com/3496491/96112842-8bc15380-0ee3-11eb-91d0-c53160d23303.png)